X5G tutorial kpm xapp (#15)
open-5g/openrangym: _posts/tutorials/2024-05-18-x5g-kpm-xapp.md

---
layout: post

> title: "KPM xApp with OSC Near-RT RIC and X5G"

Reviewer: ```suggestion
title: "KPM xApp with the OSC Near-RT RIC and NVIDIA ARC on X5G"
```

date: 2024-05-18
category: tutorials
author:

> short-description: How to run a KPM xApp leveraging the X5G testbed comprising OAI, NVIDIA ARC-OTA, and OSC Near-RT RIC

Reviewer: ```suggestion
short-description: How to run a KPM xApp leveraging the X5G testbed comprising OAI, NVIDIA ARC-OTA, and the OSC Near-RT RIC
```

---

{% assign publications = site.data.publications %}

{% assign el = publications.publications | where: "name", "villa2024x5g" %}
{% assign element = el[0] %}
{% capture pub_villa2024x5g %}
{% include pub-template.html %}
{% endcapture %}

> [X5G](https://x5g.org/) comprises of an open, programmable, and multi-vendor private 5G O-RAN testbed deployed at Northeastern University, Boston, MA, USA. It features an 8-node network deployment of the [NVIDIA Aerial RAN CoLab Over-The-Air (ARC-OTA)](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/index.html), with [OpenAirInterface](https://openairinterface.org/) for the higher layers, the [NVIDIA Aerial](https://docs.nvidia.com/aerial/index.html) Software Development Kit (SDK) for the High-PHY layer, accelerated on GPUs, and Radio Units (RUs) manufactured by various vendors, such as Foxconn. Additionally, the Near-RT RIC, developed by OSC and deployed on an OpenShift cluster, is being integrated with X5G.

Reviewer: ```suggestion
[X5G](https://x5g.org/) comprises of an open, programmable, and multi-vendor private 5G O-RAN testbed deployed at Northeastern University, Boston, MA, USA. It features an 8-node network deployment of the [NVIDIA Aerial RAN CoLab Over-The-Air (ARC-OTA)](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/index.html), with [OpenAirInterface](https://openairinterface.org/) for the higher layers, the [NVIDIA Aerial](https://docs.nvidia.com/aerial/index.html) Software Development Kit (SDK) for the High-PHY layer, accelerated on GPUs, and Radio Units (RUs) manufactured by various vendors, including Foxconn. Additionally, X5G features the O-RAN Software Community (OSC) Near-RT RIC deployed on an OpenShift cluster and integrated with an E2 agent in OAI.
```

> [X5G](https://x5g.org/) comprises of an open, programmable, and multi-vendor private 5G O-RAN testbed deployed at Northeastern University, Boston, MA, USA. It features an 8-node network deployment of the [NVIDIA Aerial RAN CoLab Over-The-Air (ARC-OTA)](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/index.html), with [OpenAirInterface](https://openairinterface.org/) for the higher layers, the [NVIDIA Aerial](https://docs.nvidia.com/aerial/index.html) Software Development Kit (SDK) for the High-PHY layer, accelerated on GPUs, and Radio Units (RUs) manufactured by various vendors, including Foxconn. Additionally, X5G features the O-RAN Software Community (OSC) Near-RT RIC deployed on an OpenShift cluster and integrated with an E2 agent in OAI.

Reviewer 2: This page does not exist

Author: This link should exist. I'll keep an eye on it.

Author: It still doesn't exist. I will merge and tell Anupa that it is not available.

Reviewer 2: Please remove the link if it does not exist. You can add it at a later stage

Author: Link removed for now.

More details can be found in the following paper:

> > {{ pub_villa2024x5g | strip_newlines }}

Reviewer 3: Unrelated to this PR, but let's maybe open an issue.
The new UI for publication does not highlight it well and should open the links in a blank page instead of substituting the actual one.

@leonardobonati where is the source code for the style of this? There are some issues that need an hotfix

Author: Which articles and links are you referring to in particular that you think it is better to remove?

Author: I will add an overview of X5G in the Platform tab of the website. Here is mostly to give people useful pointers to run the demo.

Reviewer 3: Everything that is not strictly related to the tutorial must be removed and put on the aforementioned page.

Author: Added the X5G page in the platform tab.

> {: .text-justify}

<img src="/assets/post-assets/x5g-e2e-overview.png" class="post-image" alt="X5G End-to-end Overview" width="75%">

This tutorial servers as a reference on how to operate a Key Performance Measurement (KPM) xApp in an end-to-end platform with similar capabilities to those provided by X5G. A possible final outcome of this tutorial can be found on this [X5G Overview Video](https://www.youtube.com/watch?v=_bYY12wuhzk&ab_channel=MichelePolese).

## Prerequisites
The following prerequisites are recommended to be met in order to properly run the tutorial.

### NVIDIA Aerial

> The NVIDIA Aerial information and installation guide is available at [this website](https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html).

Reviewer 3: ```suggestion
The NVIDIA Aerial information and installation guide is available at this [website](https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html).
```
```suggestion
The NVIDIA Aerial information and installation guide is available at [this website](https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html).
```

Author: I would prefer to keep [this website] to be consistent with the rest.

Reviewer 3: `this` in general is ugly in this context.

```suggestion
The NVIDIA Aerial information and installation guide is available at the [official website](https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html).
```

Author: Ok, fixed!

> The NVIDIA Aerial RAN CoLab Over-the-Air (ARC-OTA) information can be found at [this website](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/index.html).

Reviewer 2: This page does not exist

Author: This one should exist. Will be brought up soon probably (I'll keep an eye).

Author: Fixed.

Reviewer 2: It still cannot be found

Author: Check the latest commit, this now is: https://docs.nvidia.com/aerial/cuda-accelerated-ran/index.html

Reviewer 2: Thanks, it was still showing the old code snipped above

### OSC Near-RT RIC

> The O-RAN Software Community (OSC) Near-RT RIC Release E can be installed following [this guide](https://docs.o-ran-sc.org/projects/o-ran-sc-ric-plt-ric-dep/en/latest/installation-guides.html).

Reviewer: Is this specific to Release E?

Author: It is a general guide, you then specify the version during the installation.

Reviewer 2: k8s has networking issues when reaching e2term from external systems (at lease on this release of the RIC). Does it make sense to emphasize that we are running this on OpenShift?

### OpenAirInterface with Custom E2 Agent
{% assign publications = site.data.publications %}

{% assign el = publications.publications | where: "name", "moro2023open" %}
{% assign element = el[0] %}
{% capture pub_moro2023open %}
{% include pub-template.html %}
{% endcapture %}

The OpenAirInterface (OAI) gNB code requires the addition of a custom E2 Agent to manage E2SM (E2 Service Model) functionalities that has been developed and tested in the following publication:
> {{ pub_moro2023open | strip_newlines }}
> {: .text-justify}

> An OAI version with the Custom E2 Agent already integrated in the 2024w18 can be found at [this GitHub repository](https://github.com/wineslab/x5g-openairinterface/tree/2024w18_e2agent).

Reviewer: ```suggestion
The E2 agent is based on the OSC e2sim project. An OAI version with the custom E2 Agent already integrated in the 2024w18 can be found at [this GitHub repository](https://github.com/wineslab/x5g-openairinterface/tree/2024w18_e2agent).
```

### xApp

> A basic xApp developed in Python and easily extendable can be found at [this GitHub repository](https://github.com/wineslab/xapp-oai/tree/kpm-xapp). This repository contains a few Python examples that can subscribe to selected RAN parameters, receive periodic indication messages, and send control requests to update these parameters.

Reviewer 3: ```suggestion
A basic and easily extensible xApp developed in Python can be found at the [GitHub repository](https://github.com/wineslab/xapp-oai/tree/kpm-xapp). This repository contains some Python examples that can subscribe to selected RAN parameters, receive periodic indication messages, and send control requests to update these parameters.
```
```suggestion
A basic xApp developed in Python and easily extendable can be found at [this GitHub repository](https://github.com/wineslab/xapp-oai/tree/kpm-xapp). This repository contains a few Python examples that can subscribe to selected RAN parameters, receive periodic indication messages, and send control requests to update these parameters.
```

### xApp SM Connector
The xApp SM connector component connects the xApp with the OSC Near-RT RIC. On one side, it receives the custom SM buffers to be encapsulated in E2AP (E2 Application Protocol) messages and sent to the RIC. On the other side, it retrieves the custom SM buffers from E2AP to be sent to the xApp. The connector is part of the base xApp code and can be found at [this GitHub repository](https://github.com/wineslab/xapp-oai/tree/kpm-xapp).

### e2sim

> The e2sim is a component that can run on the same machine as the OAI gNB and has the duty to encapsulate/decapsulate the custom Service Model (SM) buffers to be sent to or received from the gNB. It communicates with the gNB via UDP sockets. The e2sim codebase can be found at [this GitHub repository](https://github.com/wineslab/ran-e2sim/tree/x5g-e2sim).

Reviewer: ```suggestion
The e2sim is a component that can run on the same server as the OAI L2/L3 and ARC PHY and has the duty to encapsulate/decapsulate the custom Service Model (SM) buffers to be sent to or received from the gNB. It communicates with the gNB via UDP sockets. The e2sim codebase can be found at [this GitHub repository](https://github.com/wineslab/ran-e2sim/tree/x5g-e2sim).
```

## Run the System

We assume that all components listed in the prerequisites, such as the gNB with NVIDIA ARC-OTA and OAI, OSC Near-RT RIC, e2sim, and xApp, have been compiled and are functioning properly. Please note that the commands and terminals used in this tutorial may vary depending on your setup.

To run the system in this tutorial, we open 5 terminals as follows:
- Terminal 1 (T1): NVIDIA L1 cuBB (Server 1 with A100 GPU)
- Terminal 2 (T2): OAI L2 (Server 1 same as cuBB)
- Terminal 3 (T3): e2sim (Server 1 same as cuBB)

> - Terminal 4 (T4): xApp connector (OpenShift xApp Pod)

Reviewer: ```suggestion
- Terminal 4 (T4): xApp connector (xApp Container in OSC RIC)
```

> - Terminal 5 (T5): xApp Python (OpenShift xApp Pod)

Reviewer: ```suggestion
- Terminal 5 (T5): xApp Python (xApp Container in OSC RIC)
```

- (Optional) Terminal 6 (T6): UEs

> In T1, we first run the `NVIDIA L1 cuBB`, for example, by using the following command (it may vary based on your ARC release and configurations, please refer to [ARC-OTA 1.2](https://docs.nvidia.com/aerial/aerial-ran-colab-ota/current/index.html)):

Reviewer 2: This page does not exist

Author: Fixed.

Reviewer 2: It still cannot be found

Author: Check the latest commit. This now is: https://docs.nvidia.com/aerial/cuda-accelerated-ran/aerial_cubb/cubb_quickstart/running_cubb-end-to-end.html

Reviewer 2: Thanks, it was still showing the old code snipped above
